plugin/trino-hive/src/main/java/io/trino/plugin/hive/fs/CachingDirectoryLister.java

```diff
@@ -93,14 +117,25 @@ private static SchemaTablePrefix parseTableName(String tableName)
             throws IOException
     {
         if (!isCacheEnabledFor(table.getSchemaTableName())) {
-            return fs.listLocatedStatus(path);
+            return delegate.list(fs, table, path);
+        }
+
+        ValueHolder cachedValueHolder;
+        try {
+            cachedValueHolder = cache.get(path, () -> new ValueHolder(delegate.list(fs, table, path)));
+        }
+        catch (ExecutionException e) {
+            Throwable throwable = e.getCause();
+            throwIfInstanceOf(throwable, IOException.class);
+            throwIfUnchecked(throwable);
+            throw new RuntimeException("Unexpected cause from delegate DirectoryLister", throwable);
         }
 
-        ValueHolder cachedValueHolder = uncheckedCacheGet(cache, path, ValueHolder::new);
-        if (cachedValueHolder.getFiles().isPresent()) {
-            return simpleRemoteIterator(cachedValueHolder.getFiles().get());
+        if (cachedValueHolder.isFullyCached()) {
+            return simpleRemoteIterator(cachedValueHolder.getFiles());
         }
-        return cachingRemoteIterator(cachedValueHolder, fs.listLocatedStatus(path), path);
```

Comment:
what if you'd change the current line to

```
cachingRemoteIterator(cachedValueHolder, deletegate.list(fs, table, path), path)
```

Wouldn't this help to keep things rather straightforward to understand ?

Reply:
it is
```
cachingRemoteIterator(cachedValueHolder, deletegate.list(fs, table, path), path)
```

Reply:
I meant keeping the existing implementation (from `master` ) as it is and just change the line  `103` to 

```
cachingRemoteIterator(cachedValueHolder, deletegate.list(fs, table, path), path)
```

(I hope I'm not mistaken, but )this should have the desired outcome without additional complexity.